### PR TITLE
Fix complex mode in dolfinx/dolfinx images

### PR DESCRIPTION
**Pull request by jorgensd**
As far as I can tell, the issue is that the `pip3 install` command creates a build directory that caches some info when we build the dolfinx real and then complex mode.

I've added a test to the docker.yml CI that picks up this issue, and suggested a new fix (removing the build directories from `dolfinx/python` after installation.

Should hopefully resolve https://github.com/FEniCS/dolfinx/issues/2201

### CODE REVIEW

**Comment by francesco-ballarin** [comment]
can we manually run the docker pipeline to test this?

**Comment by jhale** [comment]
The underlying reason for this might have been a change in pip related to switching from out-of-source to in-source builds, although I'm not certain.

Regarding the patch, I think it would be best to source own own configuration script `dolfinx-complex-mode` rather than manually setup the environment.

**Comment by jorgensd** [comment]
> The underlying reason for this might have been a change in pip related to switching from out-of-source to in-source builds, although I'm not certain.
> 
> Regarding the patch, I think it would be best to source our own configuration script `dolfinx-complex-mode` rather than manually setup the environment.

I wasnt able to make a one-line command that did that with `docker run`. If you could show me how to do that I’d be grateful.

**Comment by jhale** [comment]
The real underlying issue is that the DOLFINx Python object is linked with real mode PETSc objects, e.g.

```
	libamd.so.2 => /usr/local/petsc/linux-gnu-real-32/lib/libamd.so.2 (0x0000ffff8e3d0000)
	libscalapack.so.2.2 => /usr/local/petsc/linux-gnu-real-32/lib/libscalapack.so.2.2 (0x0000ffff8df70000)
	libsuperlu.so.5 => /usr/local/petsc/linux-gnu-real-32/lib/libsuperlu.so.5 (0x0000ffff8def0000)
```

**Comment by jhale** [comment]
I've triggered a build with this branch at [b7e508b](https://github.com/FEniCS/dolfinx/pull/2217/commits/b7e508becd518a7257ad81623919be5d61e692ca).

**Comment by jhale** [comment]
This patch hasn't fixed the issue. DOLFINx Python still seems to link to PETSc real mode libraries.

**Comment by jhale** [comment]
https://github.com/FEniCS/dolfinx/runs/6652047258?check_suite_focus=true

File: docker/Dockerfile [hunk 1]
```diff
@@ -418,16 +418,16 @@ ONBUILD RUN cd dolfinx && \
...
     cd ../ && \
     mkdir -p build-complex && \
     cd build-complex && \
     PETSC_ARCH=linux-gnu-complex-32 cmake -G Ninja -DCMAKE_INSTALL_PREFIX=/usr/local/dolfinx-complex -DCMAKE_BUILD_TYPE=${DOLFINX_CMAKE_BUILD_TYPE} -DCMAKE_CXX_FLAGS=${DOLFIN_CMAKE_CXX_FLAGS} ../cpp && \
     ninja install && \
     . /usr/local/dolfinx-complex/lib/dolfinx/dolfinx.conf && \
     cd ../python && \
-    CXXFLAGS=${DOLFINX_CMAKE_CXX_FLAGS} PETSC_ARCH=linux-gnu-complex-32 pip3 install -v --target /usr/local/dolfinx-complex/lib/python${PYTHON_VERSION}/dist-packages --no-dependencies . && \
+    CXXFLAGS=${DOLFINX_CMAKE_CXX_FLAGS} PETSC_ARCH=linux-gnu-complex-32 pip3 install -v --target /usr/local/dolfinx-complex/lib/python${PYTHON_VERSION}/dist-packages --no-dependencies --no-cache-dir . && \
     rm -rf build
```
Comment (jorgensd) [inline docker/Dockerfile:431]: Do want to remove this one as well?

**Pull request merged**